Refactored IErrorReportingService

333fred opened the pull request:
**Customer scenario**

As part of the work to enable A/B testing in 15.3, this refactors the IErrorReportingService to extract the infobar display component into a separate service that I'll be able to use. Submitting this PR now to get it reviewed while I work on the A/B test component. Tagging @dotnet/roslyn-ide @CyrusNajmabadi for review.

**Bugs this fixes:**

Infrastructure work for https://github.com/dotnet/roslyn/issues/19212. Not the full fix, just a part.

**Workarounds, if any**

N/A

**Risk**

Should be low. No functionality was changed by this, I just extracted the logic that was in the `VisualStudioErrorReportingService` into a new `VisualStudioInfoBarService`.

**Performance impact**

Should be small. There will be an additional object allocated and an additional method call, but they should be trivial.

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

N/A

**How was the bug found?**

Analyzer design team

333fred commented on src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioInfoBarService.cs:
```diff
@@ -0,0 +1,200 @@
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation
+{
+    [Export(typeof(IInfoBarService))]
```
@CyrusNajmabadi I made this a standard MEF component, do you want me to make it a workspace service with an `IWorkspaceServiceFactory`?

CyrusNajmabadi replied:
Yes. this shoudl be an IWorkspaceService (note: you don't need hte Factory unless you need to get at actual Workspace info).

jasonmalinowski replied:
Given you have `internal interface IInfoBarService : IWorkspaceService` it means you're expecting this to be a workspace service.

jinujoseph commented:
cc @dotnet/roslyn-ide

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/Workspaces/EditorInfoBarService.cs:
```diff
@@ -0,0 +1,23 @@
+﻿using System;
```
i would just make a top level folder called 'InfoBar' for this stuff.  I don't see why this is under Workspaces.

333fred replied:
I put it in the same folder as the EditorErrorReportingService

CyrusNajmabadi replied:
Don't :)

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/Workspaces/EditorInfoBarService.cs:
```diff
@@ -0,0 +1,23 @@
+    {
+        public void ShowInfoBarInActiveView(string message, params InfoBarUI[] items)
+        {
+            ShowInfoBarInGlobalView(message, items);
```
nit: ```=>```

CyrusNajmabadi replied:
This type doesn't seem to be exported.  How does it work?

333fred replied:
Because I missed that ErrorErrorReportingService was exported as a factory from EditorErrorReportingServiceFactory when I was copying it.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioInfoBarService.cs:
```diff
@@ -0,0 +1,200 @@
+
+        public void ShowInfoBarInActiveView(string message, params InfoBarUI[] items)
+        {
+            ShowInfoBar(activeView: true, message: message, items: items);
```
nit =>

CyrusNajmabadi replied:
add assert that this can be called from any thread (available on ForegroundThreadAffinitizedObject).

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioInfoBarService.cs:
```diff
@@ -0,0 +1,200 @@
+namespace Microsoft.VisualStudio.LanguageServices.Implementation
+{
+    [Export(typeof(IInfoBarService))]
+    internal class VisualStudioInfoBarService : IInfoBarService
```
Derive from ThreadAffinitizedObjecct

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioInfoBarService.cs:
```diff
@@ -0,0 +1,200 @@
+
+        private bool TryGetInfoBarData(bool activeView, out IVsInfoBarHost infoBarHost)
+        {
+            infoBarHost = null;
```
Add threading assert.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/ExtensionManager/IInfoBarService.cs:
```diff
@@ -0,0 +1,52 @@
+        void ShowInfoBarInGlobalView(string message, params InfoBarUI[] items);
+    }
+
+    internal struct InfoBarUI
```
struct feels overkill here.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioInfoBarService.cs:
```diff
@@ -0,0 +1,200 @@
+
+            if (activeView)
+            {
+                var monitorSelectionService = _serviceProvider.GetService(typeof(SVsShellMonitorSelection)) as IVsMonitorSelection;
```
Is all this logic just moved from the other class?

333fred replied:
Yes. All this logic was extracted from VisualStudioEditorErrorReportingService.

jasonmalinowski commented on src/Workspaces/Core/Portable/ExtensionManager/IInfoBarService.cs:
```diff
@@ -0,0 +1,52 @@
+        /// <summary>
+        /// Show global info bar
+        /// </summary>
+        /// <param name="message"></param>
```
Either document or delete.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/EditorLayerExtensionManager.cs:
```diff
@@ -70,10 +70,10 @@ public override void HandleException(object provider, Exception exception)
-                            new ErrorReportingUI(String.Empty, ErrorReportingUI.UIKind.Close, () => LogLeaveDisabled(provider)));
+                            new InfoBarUI(WorkspacesResources.Show_Stack_Trace, InfoBarUI.UIKind.HyperLink, () => ShowDetailedErrorInfo(exception), closeAfterAction: false),
+                            new InfoBarUI(WorkspacesResources.Enable, InfoBarUI.UIKind.Button, () => { EnableProvider(provider); LogEnableProvider(provider); }),
+                            new InfoBarUI(WorkspacesResources.Enable_and_ignore_future_errors, InfoBarUI.UIKind.Button, () => { EnableProvider(provider); LogEnableProvider(provider); }),
```
ew.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioInfoBarService.cs:
```diff
@@ -0,0 +1,200 @@
+        }
+
+        public void ShowInfoBarInGlobalView(string message, params InfoBarUI[] items)
+        {
```
add assert that this can be called from any thread (available on ForegroundThreadAffinitizedObject).

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioInfoBarService.cs:
```diff
@@ -0,0 +1,200 @@
+                        actionItems.Add(new InfoBarHyperlink(item.Title));
+                        break;
+                    case InfoBarUI.UIKind.Close:
+                        break;
```
??

333fred replied:
I'm not sure, this is what the existing logic does.

CyrusNajmabadi replied:
bleagh.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioInfoBarService.cs:
```diff
@@ -0,0 +1,200 @@
+                // run given onClose action if there is one.
+                items.FirstOrDefault(i => i.Kind == InfoBarUI.UIKind.Close).Action?.Invoke();
+
+                if (infoBarCookie.HasValue)
```
how could this ever not have a value?

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioInfoBarService.cs:
```diff
@@ -0,0 +1,200 @@
+            var eventSink = new InfoBarEvents(items, () =>
+            {
+                // run given onClose action if there is one.
+                items.FirstOrDefault(i => i.Kind == InfoBarUI.UIKind.Close).Action?.Invoke();
```
FirstOrDefault can reutrn null, but you dot off to get Action.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioInfoBarService.cs:
```diff
@@ -0,0 +1,200 @@
+            public void OnActionItemClicked(IVsInfoBarUIElement infoBarUIElement, IVsInfoBarActionItem actionItem)
+            {
+                var item = _items.FirstOrDefault(i => i.Title == actionItem.Text);
+                if (item.IsDefault)
```
how would this happen?

333fred commented:
@CyrusNajmabadi @jasonmalinowski addressed comments that aren't on the code I moved directly from the old implementation. Cyrus and I talked offline, we're going to leave that unchanged from the existing implementation.

333fred commented:
@dotnet/roslyn-infrastructure windows_release_vs-integration_prtest hung. Going to trigger a retest, pr build link is https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_release_vs-integration_prtest/3918/console.

333fred commented:
@dotnet-bot retest windows_release_vs-integration_prtest please

333fred commented:
Ping @CyrusNajmabadi

333fred commented:
Tagging @MattGertz for approval pending @CyrusNajmabadi's sign off.

MattGertz commented:
Approved pending signoff.

CyrusNajmabadi commented:
Looking.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/InfoBar/EditorInfoBarService.cs:
```diff
@@ -0,0 +1,22 @@
+    {
+        public void ShowInfoBarInActiveView(string message, params InfoBarUI[] items)
+        {
+            ShowInfoBarInGlobalView(message, items);
```
nit: use =>   (=> on the next line).

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/InfoBar/EditorInfoBarService.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.Workspaces
+{
+    [ExportWorkspaceService(typeof(IInfoBarService))]
```
I'm surprised there was no error on this.  I think you need to add ", Shared" to the attribute, and suck in "using System.Composition".

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/InfoBar/VisualStudioInfoBarService.cs:
```diff
@@ -0,0 +1,206 @@
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation
+{
+    [ExportWorkspaceService(typeof(IInfoBarService))]
```
Same issue with "Shared" here.  Also, you need to export this service at the "Host" layer (it's hte next parameter).  There's a constant somewhere for "Host".

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/InfoBar/VisualStudioInfoBarService.cs:
```diff
@@ -0,0 +1,206 @@
+                {
+                    CreateInfoBar(infoBarHost, message, items);
+                }
+            }, _listener.BeginAsyncOperation("Show InfoBar"));
```
use nameof(ShowInfoBar)

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioErrorReportingServiceFactory.cs:
```diff
@@ -1,34 +1,28 @@
+            {
+                _singleton = new VisualStudioErrorReportingService(workspaceServices.GetRequiredService<IInfoBarService>());
+            }
             return _singleton;
```
newline after }

CyrusNajmabadi commented:
LGTM with the changes i asked for.

CyrusNajmabadi commented:
👍
